### PR TITLE
APICompat & GenAPI clean-up

### PR DESCRIPTION
**Pull request by ViktorHofer**
Now that we stabilized APICompat & GenAPI for .NET 8, I wanted to reduce code size and simplify things.

- Use new language features (primary constructors, collection initializers)
- Analyzers (prefer concrete return type, use local methods)
- Don't use triple slash XML comments for non public API (otherwise those would be emitted into the satellite assemblies)
- Use `HashCode.Combine` on .NET
- Fix typos
- GenAPIApp code simplification

### CODE REVIEW

File: src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/ApiCompatServiceProvider.cs [hunk 1]
```diff
@@ -17,7 +17,7 @@ internal sealed class ApiCompatServiceProvider
         private readonly Lazy<ISuppressibleLog> _compatibilityLogger;
         private readonly Lazy<IApiCompatRunner> _apiCompatRunner;
 
-        internal ApiCompatServiceProvider(Func<ISuppressionEngine, ISuppressibleLog> logFactory,
+        public ApiCompatServiceProvider(Func<ISuppressionEngine, ISuppressibleLog> logFactory,
```
Comment (carlossanlop) [inline src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/ApiCompatServiceProvider.cs:20]: Why did you decide to make this `public`? I don't see new invocations.

File: src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 2]
```diff
@@ -12,11 +12,9 @@ namespace Microsoft.DotNet.GenAPI
...
     {
-        /// <summary>
-        /// Creates a declaration matching an existing symbol.
-        ///     The reason of having this similar to `SyntaxGenerator.Declaration` extension method is that
-        ///     SyntaxGenerator does not generates attributes neither for types, neither for members.
-        /// </summary>
+        // Creates a declaration matching an existing symbol.
+        // The reason of having this similar to `SyntaxGenerator.Declaration` extension method is that
+        // SyntaxGenerator does not generates attributes neither for types, neither for members.
         public static SyntaxNode DeclarationExt(this SyntaxGenerator syntaxGenerator, ISymbol symbol, ISymbolFilter symbolFilter)
```
Comment (carlossanlop) [inline src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:18]: nit: The `*Ext` suffix in this extension method name is redundant/unnecessary. If it's not too disruptive, consider removing it. I am not sure if the name `Declaration` would cause issues.

File: src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/IMethodSymbolExtensions.cs [hunk 3]
```diff
@@ -8,14 +8,9 @@ namespace Microsoft.DotNet.GenAPI
...
-        /// <summary>
-        /// Compare the equality of two method signatures for the purpose of emitting a "new"
-        /// keyword on a method's return type. This is *not* meant to be complete implementation,
-        /// but rather a heuristic to check that one method may hide another.
-        /// </summary>
-        /// <param name="method"></param>
-        /// <param name="otherMethod"></param>
-        /// <returns>true if the signatures are determined to be equal</returns>
+        // Compare the equality of two method signatures for the purpose of emitting a "new"
+        // keyword on a method's return type. This is *not* meant to be complete implementation,
```
Comment (carlossanlop) [inline src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/IMethodSymbolExtensions.cs:12]:
```suggestion
        // keyword on a method's return type. This is *not* meant to be a complete implementation,
```

File: src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/GenAPIApp.cs [hunk 4]
```diff
@@ -165,13 +103,7 @@ public static void Run(ILog logger, Context context)
...
         }
 
-        /// <summary>
-        /// Creates a TextWriter capable to write into Console or cs file.
-        /// </summary>
-        /// <param name="outputDirPath">Path to a directory where file with `assemblyName`.cs filename needs to be created.
-        ///     If Null - output to Console.Out.</param>
-        /// <param name="assemblyName">Name of an assembly. if outputDirPath is not a Null - represents a file name.</param>
-        /// <returns></returns>
+        // Creates a TextWriter capable to write into Console or cs file.
```
Comment (carlossanlop) [inline src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/GenAPIApp.cs:106]:
```suggestion
        // Creates a TextWriter capable of writing into Console or a cs file.
```

File: src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Runner/ApiCompatRunner.cs [hunk 5]
```diff
@@ -13,44 +13,32 @@ namespace Microsoft.DotNet.ApiCompatibility.Runner
...
         {
-            _log.LogMessage(MessageImportance.Low,
+            log.LogMessage(MessageImportance.Low,
                 string.Format(Resources.ApiCompatRunnerExecutingWorkItems,
                     _workItems.Count));
 
             foreach (ApiCompatRunnerWorkItem workItem in _workItems)
             {
-                IReadOnlyList<ElementContainer<IAssemblySymbol>> leftContainerList = CreateAssemblySymbols(workItem.Left, workItem.Options, out bool resolvedExternallyProvidedAssemblyReferences);
+                List<ElementContainer<IAssemblySymbol>> leftContainerList = CreateAssemblySymbols(workItem.Left, out bool resolvedExternallyProvidedAssemblyReferences);
```
Comment (carlossanlop) [inline src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Runner/ApiCompatRunner.cs:35]: Why did you change them to List?

File: src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Mapping/TypeMapper.cs [hunk 6]
```diff
@@ -13,35 +13,25 @@ namespace Microsoft.DotNet.ApiCompatibility.Mapping
     /// This also holds the nested types as a list of <see cref="ITypeMapper"/> and the members defined within the type
     /// as a list of <see cref="IMemberMapper"/>
     /// </summary>
-    public class TypeMapper : ElementMapper<ITypeSymbol>, ITypeMapper
+    /// <param name="ruleRunner">The <see cref="IRuleRunner"/> that compares the type mapper elements.</param>
+    /// <param name="settings">The <see cref="IMapperSettings"/> used to compare the type mapper elements.</param>
+    /// <param name="rightSetSize">The number of elements in the right set to compare.</param>
+    /// <param name="containingNamespace">The containing <see cref="INamespaceMapper"/>.</param>
+    /// <param name="containingType">The containing <see cref="ITypeMapper"/>. Null, if the type doesn't have a containing type.</param>
```
Comment (carlossanlop) [inline src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Mapping/TypeMapper.cs:20]: nit
```suggestion
    /// <param name="containingType">The containing <see cref="ITypeMapper"/>. <see langword="null" />, if the type doesn't have a containing type.</param>
```

File: src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Mapping/AssemblyMapper.cs [hunk 7]
```diff
@@ -11,33 +11,24 @@ namespace Microsoft.DotNet.ApiCompatibility.Mapping
     /// This also holds a list of <see cref="INamespaceMapper"/> to represent the mapping of namespaces in between
     /// <see cref="IElementMapper{T}.Left"/> and <see cref="IElementMapper{T}.Right"/>.
     /// </summary>
-    public class AssemblyMapper : ElementMapper<ElementContainer<IAssemblySymbol>>, IAssemblyMapper
+    /// <param name="ruleRunner">The <see cref="IRuleRunner"/> that compares the assembly mapper elements.</param>
+    /// <param name="settings">The <see cref="IMapperSettings"/> used to compare the assembly mapper elements.</param>
+    /// <param name="rightSetSize">The number of elements in the right set to compare.</param>
+    /// <param name="containingAssemblySet">The containing <see cref="IAssemblySetMapper"/>. Null, if the assembly isn't part of a set.</param>
```
Comment (carlossanlop) [inline src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Mapping/AssemblyMapper.cs:17]: nit
```suggestion
    /// <param name="containingAssemblySet">The containing <see cref="IAssemblySetMapper"/>. <see langword="null" />, if the assembly isn't part of a set.</param>
```

File: src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs [hunk 8]
```diff
@@ -10,33 +10,24 @@ namespace Microsoft.DotNet.ApiCompatibility.Logging
...
+    /// <param name="baselineAllErrors">If true, baselines all errors.</param>
+    public class SuppressionEngine(string? noWarn = null, bool baselineAllErrors = false) : ISuppressionEngine
     {
         protected const string DiagnosticIdDocumentationComment = " https://learn.microsoft.com/en-us/dotnet/fundamentals/package-validation/diagnostic-ids ";
-        private readonly HashSet<Suppression> _baselineSuppressions = new();
-        private readonly HashSet<Suppression> _suppressions = new();
-        private readonly HashSet<string> _noWarn;
+        private readonly HashSet<Suppression> _baselineSuppressions = [];
+        private readonly HashSet<Suppression> _suppressions = [];
+        private readonly HashSet<string> _noWarn = string.IsNullOrEmpty(noWarn) ? [] : new HashSet<string>(noWarn!.Split(';'));
```
Comment (carlossanlop) [inline src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs:20]: I'm surprised you had to use `noWarn!`. I would've assumed the nullability feature would know you already checked it.

File: src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/ApiCompatServiceProvider.cs [hunk 9]
```diff
@@ -17,7 +17,7 @@ internal sealed class ApiCompatServiceProvider
         private readonly Lazy<ISuppressibleLog> _compatibilityLogger;
         private readonly Lazy<IApiCompatRunner> _apiCompatRunner;
```
Comment (carlossanlop) [inline src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/ApiCompatServiceProvider.cs:19]: Maybe sprinkle some triple slash comments if this is public now.

**Review by carlossanlop — APPROVED** [review]
LGTM. Nothing blocking, but left some questions as well as some suggestions for you to consider.

File: src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/ApiCompatServiceProvider.cs (hunk 1 above)
Comment (ViktorHofer) [reply src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/ApiCompatServiceProvider.cs:20]: The class is already internal and I prefer to have public API in an internal type over internal API in an internal type.

File: src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Runner/ApiCompatRunner.cs (hunk 5 above)
Comment (ViktorHofer) [reply src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Runner/ApiCompatRunner.cs:35]: An analyzer suggested that: prefer concrete return type

File: src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/ApiCompatServiceProvider.cs (hunk 9 above)
Comment (ViktorHofer) [reply src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompat.Shared/ApiCompatServiceProvider.cs:19]: The ApiCompatServiceProvider type is internal so no need for triple slash comments. They are actually harmful as they still get embedded into the compiler generated XML file which we don't want for non public API.

File: src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs (hunk 8 above)
Comment (ViktorHofer) [reply src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs:20]: Yeah the `!` isn't necessary on .NETCoreApp but on .NETFramework which has inferior nullability compiler heuristics.

File: src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 2 above)
Comment (andriipatsula) [reply src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:18]: I'm not entirely certain, but I believe my wanted to highlight that the method is indeed an extension method. Feel free to rename it.
